driver: sensor: max30101 Enable multiple instance, data acquisition and add trigger support

Sokario opened the pull request:
This pull request adds multiple instance, triggers support and enhance data recovery for the MAXIM MAX30101.

**MAXIM MAX30101 sensor multi-instance support**:

- Removed instance based Kconfig to device tree bindings, added parameter's recuperation from device tree.
- More readable parameters for device tree.
- Added missing configuration registers.
- Updated nodes using max30101 compatible:
  - **hexiwear_mk64f12** board (boards/nxp/hexiwear/hexiwear_mk64f12.dts)
  - **i2c.dtsi** tests drivers (test/drivers/build_all/sensor/i2c.dtsi)

**MAXIM MAX30101 sensor triggers support**:

- Added new **SENSOR_CHAN** and **SENSOR_TRIG**:
  - **SENSOR_CHAN_AMBIENT_LIGHT**: because MAX30101 allows verification of ambient light cancellation.
  - **SENSOR_TRIG_OVERFLOW**: the MAX30101 allows interrupt to be raised when ambient light cancellation overflows.
- Added interrupt related Kconfig options and `irq-gpios` device tree parameter.
- Added `max30101_trigger.c` and updated `CMakeLists.txt` and `max30101.c`.
- Added interrupt support in samples: **samples/sensor/heart_rate**
- Updated `sample.yaml` and `README.rst` accordingly.
- Updated `sensor_shell.c` and `test_sensor_shell.py` to match the tow newly added enumerators.
- Updated `test_sensor_shell.py` to automatically adapt to futur modification of **sensor_channel**.

**MAXIM MAX30101 sensor `sample_fetch` enhancement**:

- Added better data recovery for more complex multi-LED configurations.
- New fifo/channel matrix mapping.
- Added **MAX30101_DIE_TEMPERATURE** Kconfig to add die temperature acquisition (made it Kconfig in order to let the user remove the extra communication time on tight period acquisition).
- Added Kconfig condition on temperature triggers.

MAX30101 [datasheet](https://www.analog.com/media/en/technical-documentation/data-sheets/max30101.pdf).

Sokario committed:
drivers: sensor: max30101: Enable support for multiple instance

The max30101 sensor driver doesn't support multiple instance.
Update Kconfig and maxim,max30101.yaml for instance based
configuration. Propagate changes over existing files.

Signed-off-by: Logan Saint-Germain <l.saintgermain@catie.fr>

Sokario committed:
drivers: sensor: max30101: Enhanced driver to support triggers

The max30101 sensor driver doesn't support triggers.
Add `.trigger_set` API and corresponding Kconfig and
device tree parameters. Add `SENSOR_CHAN_AMBIENT_LIGHT`
and `SENSOR_TRIG_OVERFLOW`.

Signed-off-by: Logan Saint-Germain <l.saintgermain@catie.fr>

MaureenHelm reviewed (changes requested):
Any chance you're planning to implement the sensor async API?

Sokario committed:
drivers: sensor: max30101: Enhanced sample_fetch to match datasheet

The max30101 allows to configure time slots in samples acquisition.
It is now supported by adding matrix mapping for the slot/fifo
indexing. When a channel is present multiple times, the resulting
sample from the `sensor_channel_get` is averaging each entry.
Added Die temperature sample acquisition with
`CONFIG_MAX30101_DIS_TEMPERATURE` Kconfig.

Signed-off-by: Logan Saint-Germain <l.saintgermain@catie.fr>

Sokario commented:
> Any chance you're planning to implement the sensor async API?

I wasn't planning to do it because it didn't know it existed, but I will be working on it from now on.
However, it will be in another PR.

I integrated your review and added another commit as I discovered the drivers wasn't taking care of each possible configuration in the multi-LED mode. It is now possible to give multiple times the colour channel chosen.
However, it results in multiple sample in one acquisition, which break the zephyr sensor API. In order to tackle this issue, the resulting sample of `sensor_channel_get` is the average of all the acquired samples.

I am updating the PR name and description to match the modifications.

Sokario commented:
I had to modify once again the `test_sensor_shell.py` because of newly added commits which added **flow rate**. Doesn't twister warn about it ?
I changed the fixed channel number in `test_sensor_shell_attr_set/get` for its name as it is supported in `sensor_shell.c`. It should allows anyone to add new channel whithout breaking the attr setter and getter of sensor_shell pytests.

Sokario committed:
samples: sensor: sensor_shell: Added channels count for sensor_shell tests

Use of the `sensor get` help. No channel provided allows the sensor_shell
to iterate through every channels. Getting the last channel gives the last
channel index and therefor the channel count. Provide futur proofing for
new channels.

Signed-off-by: Logan Saint-Germain <l.saintgermain@catie.fr>

Sokario commented:
While working on async API, I noticed **SENSOR_TRIG_FIFO_FULL** wasn't the best suited for `FIFO_A_FULL` interruption after all. I switched for  **SENSOR_TRIG_FIFO_WATERMARK** wich better correspond to what `FIFO_A_FULL` interruption do (It allows to trigger to a set number of sample acquired in the max30101 fifo).

Sokario commented:
Thanks everyone for the reviews, feedback, and quick actions!
I really appreciate how efficiently the discussions and updates helped move this PR forward.
It’s a pleasure to contribute 👍
